File: rust/2022/src/bin/11.rs
// https://adventofcode.com/2022/day/11

use regex::Regex;

#[derive(Debug)]
enum Operand {
    Number(u32),
    Old,
}

#[derive(Debug)]
struct Monkey {
    operation: (char, Operand),
    test: u32,
    if_true: usize,
    if_false: usize,
    inspection_count: u32,
}

impl Monkey {
    fn new(operation: (char, Operand), test: u32, if_true: usize, if_false: usize) -> Monkey {
        Monkey {
            operation,
            test,
            if_true,
            if_false,
            inspection_count: 0,
        }
    }
}

fn parse_input(input: &str) -> (Vec<Monkey>, Vec<Vec<u32>>) {
    let re = Regex::new(r"old (\+|\*) (.+)").unwrap();

    let mut monkeys: Vec<Monkey> = vec![];
    let mut bags: Vec<Vec<u32>> = vec![];

    input.split("\n\n").for_each(|group| {
        let lines = group.lines().collect::<Vec<&str>>();

        let items = lines[1]
            .split(": ")
            .nth(1)
            .unwrap()
            .split(", ")
            .map(|num| num.parse::<u32>().unwrap())
            .collect::<Vec<u32>>();

        bags.push(items);

        let caps = re.captures(lines[2]).unwrap();
        let operand = match &caps[2] {
            "old" => Operand::Old,
            num => Operand::Number(num.parse().unwrap()),
        };
        let operation = (caps[1].chars().nth(0).unwrap(), operand);

        let test = lines[3].split_whitespace().last().unwrap().parse().unwrap();
        let if_true = lines[4].split_whitespace().last().unwrap().parse().unwrap();
        let if_false = lines[5].split_whitespace().last().unwrap().parse().unwrap();

        monkeys.push(Monkey::new(operation, test, if_true, if_false));
    });

    (monkeys, bags)
}

fn do_round(monkeys: &mut Vec<Monkey>, bags: &mut Vec<Vec<u32>>) {
    // During each round, give each monkey a turn
    monkeys.iter_mut().enumerate().for_each(|(i, monkey)| {
        let bag = bags[i].clone();
        bags[i].clear();

        bag.iter().for_each(|item| {
            // Inspect item
            let mut new_item = match monkey.operation {
                ('+', Operand::Number(num)) => item + num,
                ('+', Operand::Old) => item + item,
                ('*', Operand::Number(num)) => item * num,
                ('*', Operand::Old) => item * item,
                _ => unreachable!(),
            };

            // Post-inspection worry reduction
            new_item /= 3;

            // Decide which monkey to throw the item to
            if new_item % monkey.test == 0 {
                bags[monkey.if_true].push(new_item);
            } else {
                bags[monkey.if_false].push(new_item);
            };

            // Increment inspection count
            monkey.inspection_count += 1;
        });
    });
}

fn part1(input: &str) -> u32 {
    let (mut monkeys, mut bags) = parse_input(input);

    for _ in 0..20 {
        do_round(&mut monkeys, &mut bags);
    }

    let mut inspection_counts = monkeys
        .iter()
        .map(|monkey| monkey.inspection_count)
        .collect::<Vec<_>>();

    inspection_counts.sort_by(|a, b| b.cmp(a));

    inspection_counts[0] * inspection_counts[1]
}

fn main() {
    let input = include_str!("../../../../_input/2022/11.txt").trim_end();

    println!("part 1: {}", part1(input));
}

#[cfg(test)]
mod tests {
    const INPUT: &str = include_str!("../../../../_input/2022/11.txt");

    #[test]
    fn part1() {
        assert_eq!(super::part1(INPUT.trim_end()), 151312);
    }
}

File: typescript/2022/11.ts
// https://adventofcode.com/2022/day/11

class Monkey {
  public items: number[];
  public operation: (x: number) => number;
  public test: number;
  public ifTrue: number;
  public ifFalse: number;

  public inspectionCount = 0;

  constructor(items: number[], operator: string, opValue: string, test: number, ifTrue: number, ifFalse: number) {
    this.items = items;

    const parsedOpValue = parseInt(opValue, 10);

    if (operator === "*") {
      if (isNaN(parsedOpValue)) {
        this.operation = (x: number) => x * x;
      } else {
        this.operation = (x: number) => x * parsedOpValue;
      }
    } else if (operator === "+") {
      if (isNaN(parsedOpValue)) {
        this.operation = (x: number) => x + x;
      } else {
        this.operation = (x: number) => x + parsedOpValue;
      }
    }

    this.test = test;
    this.ifTrue = ifTrue;
    this.ifFalse = ifFalse;
  }

  public inspectItems(monkeys: Monkey[], throwTo: (monkey: Monkey, item: number) => void) {
    this.inspectionCount += this.items.length;

    this.items.forEach(item => {
      // Inspect item
      let itemValue = this.operation(item);
      // Post-inspection worry reduction
      itemValue = Math.floor(itemValue / 3);

      // Decide which monkey to throw the item to
      if (itemValue % this.test === 0) {
        throwTo(monkeys[this.ifTrue], itemValue);
      } else {
        throwTo(monkeys[this.ifFalse], itemValue);
      }
    });

    this.items = [];
  }
}

function createMonkeys(input: string): Monkey[] {
  const monkeys: Monkey[] = [];

  // Group lines by empty lines
  const groupedInput = input.split("\n").reduce(
    (groups, line) => {
      line === "" ? groups.push([]) : groups[groups.length - 1].push(line);
      return groups;
    },
    [[]]
  );

  groupedInput.forEach(lines => {
    const itemsMatch = lines[1].match(/(\d+),?/g);
    const items = itemsMatch.map(item => parseInt(item, 10));

    const operationMatch = lines[2].match(/(\+|\*) (.+)/);
    const [_, operator, opValue] = operationMatch;

    const testMatch = lines[3].match(/(\d+)$/);
    const test = parseInt(testMatch[1], 10);

    const ifTrueMatch = lines[4].match(/(\d+)$/);
    const ifTrue = parseInt(ifTrueMatch[1], 10);

    const ifFalseMatch = lines[5].match(/(\d+)$/);
    const ifFalse = parseInt(ifFalseMatch[1], 10);

    monkeys.push(new Monkey(items, operator, opValue, test, ifTrue, ifFalse));
  });

  return monkeys;
}

function throwTo(monkey: Monkey, item: number): void {
  monkey.items.push(item);
}

function part1(input: string): number {
  const monkeys = createMonkeys(input);

  const roundCount = 20;

  // Run game for 20 rounds
  for (let i = 0; i < roundCount; i++) {
    // During each round, give each monkey a turn
    monkeys.forEach(monkey => {
      monkey.inspectItems(monkeys, throwTo);
    });
  }

  const inspectionCounts = monkeys
    .map(monkey => monkey.inspectionCount)
    .sort((a, b) => a - b)
    .reverse();

  return inspectionCounts[0] * inspectionCounts[1];
}

const input = (await Bun.file("../_input/2022/11.txt").text()).trimEnd();

if (import.meta.env.NODE_ENV === "test") {
  const { test, expect } = await import("bun:test");

  test("part 1", () => expect(part1(input)).toBe(151312));
} else {
  console.log("part 1:", part1(input));
}
